### PR TITLE
feat(abigen): Add flag let be able to add user tmpl file.

### PR DESCRIPTION
**Pull request by mask-pp**
## 1. Purpose or design rationale of this PR
Add flag let be able to add user tmpl file.
...

## 2. PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [ ] build: Changes that affect the build system or external dependencies (example scopes: yarn, eslint, typescript)
- [ ] ci: Changes to our CI configuration files and scripts (example scopes: vercel, github, cypress)
- [ ] docs: Documentation-only changes
- [x] feat: A new feature
- [ ] fix: A bug fix
- [ ] perf: A code change that improves performance
- [ ] refactor: A code change that doesn't fix a bug, or add a feature, or improves performance
- [ ] style: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
- [ ] test: Adding missing tests or correcting existing tests

## 3. Deployment tag versioning

Has the version in `params/version.go` been updated?

- [x] This PR doesn't involve a new deployment, git tag, docker image tag, and it doesn't affect traces
- [ ] Yes

## 4. Breaking change label

Does this PR have the `breaking-change` label?

- [x] This PR is not a breaking change
- [ ] Yes

### CODE REVIEW

File: accounts/abi/bind/template.go [hunk 1]
```diff
@@ -79,6 +79,13 @@ var tmplSource = map[Lang]string{
 	LangJava: tmplSourceJava,
 }
 
+// SetTmplSource supports this func in order to set special tmpl file.
```
Comment (0xmountaintop) [inline accounts/abi/bind/template.go]:
```suggestion
// SetTmplSource supports this func in order to set special template file.
```

File: cmd/abigen/main.go [hunk 2]
```diff
@@ -265,6 +270,15 @@ func abigen(c *cli.Context) error {
 			aliases[match[1]] = match[2]
 		}
 	}
+	// Set customize tmpl file.
```
Comment (0xmountaintop) [inline cmd/abigen/main.go]:
```suggestion
	// Set customize template file.
```

File: cmd/abigen/main.go [hunk 3]
```diff
@@ -265,6 +270,15 @@ func abigen(c *cli.Context) error {
...
+	if c.GlobalIsSet(tmplFlag.Name) {
+		tmplFile := c.GlobalString(tmplFlag.Name)
+		data, err := os.ReadFile(tmplFile)
+		if err != nil {
+			utils.Fatalf("Failed to read tmpl file: %v", err)
```
Comment (0xmountaintop) [inline cmd/abigen/main.go]:
```suggestion
			utils.Fatalf("Failed to read template file: %v", err)
```

File: cmd/abigen/main.go [hunk 4]
```diff
@@ -265,6 +270,15 @@ func abigen(c *cli.Context) error {
...
+		}
+		bind.SetTmplSource(bind.LangGo, string(data))
```
Comment (0xmountaintop) [inline cmd/abigen/main.go]: make use of `langFlag`?
like 
`bind.SetTmplSource(c.GlobalString(langFlag.Name), string(data))`?

File: cmd/abigen/main.go [hunk 5]
```diff
@@ -103,6 +103,10 @@ var (
 		Name:  "contract",
 		Usage: "Name of the contract to generate the bindings for",
 	}
+	tmplFlag = cli.StringFlag{
+		Name:  "tmpl",
+		Usage: "Support user tmpl file",
```
Comment (0xmountaintop) [inline cmd/abigen/main.go]:
```suggestion
		Usage: "Template file if a user wants to customize",
```

**Review by colinlyguo — DISMISSED** [review]

File: cmd/abigen/main.go (hunk 4 above)
Comment (mask-pp) [reply cmd/abigen/main.go]: It's ok, I will change it.

Comment (mask-pp) [reply cmd/abigen/main.go:280]: `langFlag` already exists, we should change it to use another name.

**Review by colinlyguo — APPROVED** [review]

**Review by Thegaram — COMMENTED** [review]
What are these templates used for? Can you please provide an example for how (and why) we'd use this new flag?

**Comment by mask-pp** [comment]
> What are these templates used for? Can you please provide an example for how (and why) we'd use this new flag?


It is not convenient to display here, but after adding this flag, users can easily customize their own translation template files.

**Comment by 0xmountaintop** [comment]
> > What are these templates used for? Can you please provide an example for how (and why) we'd use this new flag?
> 
> It is not convenient to display here, but after adding this flag, users can easily customize their own translation template files.

I think it would be better if you can briefly describe it here.
This also serves the need for explaining "Purpose or design rationale of this PR"

**Comment by mask-pp** [comment]
> I think it would be better if you can briefly describe it here. This also serves the need for explaining "Purpose or design rationale of this PR"

Create a tmpl file and translate to `go` source file, in this way, we can simplify the code which we need to sync with the contract.

**Example command:**

```
abigen --sol scroll/L1/gateways/XX.sol --pkg bytecode --out ./XX.go --contract XX --tmpl ./metrics.tmpl
```

**A simple tmpl file**: 
```
// Code generated - DO NOT EDIT.
// This file is a generated binding and any manual changes will be lost.

package {{.Package}}

import (
	"math/big"
	"strings"
	"errors"

	ethereum "github.com/scroll-tech/go-ethereum"
	"github.com/scroll-tech/go-ethereum/accounts/abi"
	"github.com/scroll-tech/go-ethereum/accounts/abi/bind"
	"github.com/scroll-tech/go-ethereum/common"
	"github.com/scroll-tech/go-ethereum/core/types"
	"github.com/scroll-tech/go-ethereum/event"
)

// Reference imports to suppress errors if they are not otherwise used.
var (
	_ = errors.New
	_ = big.NewInt
	_ = strings.NewReader
	_ = ethereum.NotFound
	_ = bind.Bind
	_ = common.Big1
	_ = types.BloomLookup
	_ = event.NewSubscription
)

{{$structs := .Structs}}
{{range $structs}}
	// {{.Name}} is an auto generated low-level Go binding around an user-defined struct.
	type {{.Name}} struct {
	{{range $field := .Fields}}
	{{$field.Name}} {{$field.Type}}{{end}}
	}
{{end}}
```

**Review by Thegaram — APPROVED** [review]

File: accounts/abi/bind/template.go [hunk 6]
```diff
@@ -79,6 +79,13 @@ var tmplSource = map[Lang]string{
 	LangJava: tmplSourceJava,
 }
 
+// SetTmplSource supports this func in order to set special template file.
+func SetTmplSource(lang Lang, source string) {
+	if _, ok := tmplSource[lang]; ok {
```
Comment (0xmountaintop) [inline accounts/abi/bind/template.go:84]: do we need this check?

Comment (mask-pp) [reply accounts/abi/bind/template.go:84]: We need to check it, `abigen` can't translate the unsupported language.

**Pull request merged**